servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/ClientSslConfig.java

```diff
@@ -39,14 +39,23 @@
 
     /**
      * Get the non-authoritative name of the peer.
-     * @return the non-authoritative name of the peer, or {@code null} if unavailable (which may disable
-     * {@link #hostnameVerificationAlgorithm() hostname verification} and
+     * @return the non-authoritative name of the peer, or {@code null} if unavailable (which may require disabling
+     * {@link #hostnameVerificationAlgorithm()} hostname verification, and may disable
      * <a href="https://tools.ietf.org/html/rfc5077">session resumption</a>).
      * @see SSLEngine#getPeerHost()
      */
     @Nullable
     String peerHost();
 
+    /**
+     * Determine if the user explicitly set the {@link #peerHost()}.
+     *
+     * @return {@code true} the user explicitly set {@link #peerHost()} to some value (including {@code null}).
+     */
+    default boolean isPeerHostSet() {
```

Comment:
1. `ClientSslConfig` is a relatively new API and we do not expect people implementing their own version. Not necessary to make it `default`.
2. Instead of introducing a new API method, can we consider using an empty string as a way to reset the peerHost?

@Scottmitch `SSLEngine` takes `peerHost` and `peerPort` together. Imho, our API should also keep them as one method that takes/returns `HostAndPort`. Otherwise, it's easier to make a mistake of setting one, but not another. Or at least have two variants:
```java
public ClientSslConfigBuilder peerHost(String);
public ClientSslConfigBuilder peerHostAndPort(HostAndPort);
```
What can be a use-case when users want to override the `peerPort`, but not the `peerHost`?

Reply:
Great catch, @idelpivnitskiy. I think so far everything turned out fine for HTTP. The host AND port inference from target address would not have been disabled even when the user set `peerHost` to something of their desire. In that case the host would have changed, but the port would have been inferred. In the implementation I suggested, not setting the port would result in using `-1` even if multiple ports were used for connections, which would disable SSL caching of sessions ([check JDK implementation](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.base/share/classes/sun/security/ssl/SSLSessionContextImpl.java#L178)).

Therefore, I agree with your suggestion for creating a method `peerHostAndPort(HostAndPort)`. This would make the behaviour easier to understand from the user's perspective.

Let's assume an empty string represents the situation when the user explicitly set `peerHost` to be empty.

I would imagine, that setting it to empty also disables port inference (effectively, as shown by JDK implementation, if peerHost is not set then the port is not used for caching either. For hostname verification, only the peerHost/sni is used.

The questions that arise are:
* if the user set `peerHost(String)` to non-empty but has not fiddled with `peerPort`, do we keep inferring the port?
* if the user explicitly wants to set `peerPort(int)` to some disabled value to disable the inference – what value would that be?

With existing API for `peerHost` the `String` type allows three independent value types: `null`, `""` and non-empty strings. For `int` it is different, we could potentially explain `-1` stands for "go ahead and infer", while `< -1` stands for "use -1 and don't infer the port". I feel the interface would be getting clunky at this point.

---

There is another way to approach the problem, as suggested by @Scottmitch – introducing new API methods in `SingleAddressHttpClientBuilder` to disable inference of `peerHost`/`peerPort`, and possibly disabling `sniHostname` inference and SNI use. However, I think that those settings are integral to `ClientSslConfig` regardless of the protocol (currently there's only HTTP, but potentially raw TCP/UDP could utilize those settings).

Having all that in mind, I'd suggest two options for changes to `ClientSslConfig`:

**(1)**

* Add method `Supplier<HostAndPort> peerHostAndPort()`: when null -> default behaviour = infer both from target address; when returned `Supplier.get()` returns null -> disable inference = use default empty values (`null`, `-1`); when not `null` -> use whatever the user provided,
* Add method `Supplier<String> getSniHostname()`: when null -> default behaviour = infer from target address; when returned `Supplier.get()` returns null -> disable SNI; when not `null` -> use for SNI Hostname,
* Deprecate/remove `String sniHostname()`,
* Deprecate/remove `String peerHost()`,
* Deprecate/remove `int peerPort()`.

**(2)**
* Leave `String sniHostname()`, `String peerHost()`, `int peerPort()` in their current form,
* Add methods `boolean isPeerHostSet()`, `boolean isPeerPortSet()` which allow determining whether the user set the values and wants to disable any default inference for any of them,
* `isSniEnabled()` -> returns `true` by default or when user provided a non- `null`/empty value. Returns `false` when user explicitly set `null`. The builder should disallow empty strings. As a result: `isSniEnabled() && sniHostname() == null` results in address inference for SNI, `isSniEnabled() && sniHostname() != null) results in using the provided hostname and disabling SNI for `!isSniEnabled()`.

Please let me know whether any of those options work and I'll get back to the implementation.

Reply:
It seems like the `ClientSslConfig` API provides what it needs to, and the confusion is introduced at the HTTP layer which attempts to infer and apply overrides under some circumstances. If we allow the user to disable the http infer/override and just use the `ClientSslConfig` verbatim, would that address the original issue? This would avoid over complicating the `ClientSslConfig` API.

Reply:
Agreed, the implementation will be simpler. My only worry is that encapsulation of SSL configuration is lost and the user needs to interact with two builders:

- `ClientSslConfigBuilder` to set the config values,
- and `SingleAddressHttpClientBuilder` to tell the `HttpClient` what to do with the provided values.

Sample use that makes the client use SNI which is inferred, but does not verify the certificates and disables session caching:

```
        HttpClients.forSingleAddress(serverHostAndPort(serverCtx))
                .sslConfig(
                        new ClientSslConfigBuilder(DefaultTestCerts::loadServerCAPem)
                                .hostnameVerificationAlgorithm("null")
                                .build()
                )
                .disablePeerHostInference()
                .buildBlocking())
```

I think this depicts a reasonable use case. If that is fine by both of you, @Scottmitch and @idelpivnitskiy, I'll go ahead and make changes.